File: Winter-2019/linked_lists/node_value/getvalue.cpp
#include <iostream>
#include <sstream>

using namespace std;

class Node {

    public: 
        int data;
        Node* next = nullptr;

        Node(int data) {
            this->data = data;
            this->next = nullptr;
        }

        // constructor assumes that array has non-zero length
        // bad constructor! bad!
        Node(int arr[], int length) {
            this->data = arr[0];
            Node* curr = nullptr;
            Node* prev = nullptr;
            for (int i = 1; i < length; i++) {
                curr = new Node(arr[i]);
                if (i == 1) this->next = curr;
                else prev->next = curr;
                prev = curr;
            }
        }

};

void getValueRecursiveHelper(Node* head, int& k, int original, int& val) {
    if (head == nullptr) {
        k--;
        return;
    }
    getValueRecursiveHelper(head->next, k, original, val);
    if (k == 0) val += (0 - val + head->data);
    if (k < original) k--;
}

int getValue(Node* head, int& k) {
    int result = -1;
    getValueRecursiveHelper(head, k, k, result);
    return result;
}

int main() {
    int arr[] = {8, 12, -3, 4, 9, 2};
    Node* head = new Node(arr, 6);
    int v1 = 3;
    cout << "k = 3:\t" << getValue(head, v1) << "\t(Expecting 4)" << endl;
    int v2 = 8;
    cout << "k = 8:\t" << getValue(head, v2) << "\t(Expecting -1)" << endl;
    return 0;
}

File: Winter-2019/linked_lists/node_value/GetValue.java
public class GetValue {

    public static int getValueIterative(Node head, int target) {
        Node result = head;
        for (int i = 0; head != null; i++) {
            head = head.next;
            // preserve the interval between head and result
            if (i >= target) result = result.next;
        }
        return result.data;
    }

    public static void print(Node head) {
        if (head == null) return;
        System.out.print(head.data + " ");
        print(head.next);
    }

    public static void main(String[] args) {
        int[] arr = { 3, 18, 4, 9, 12, -6, 2 };
        Node head = new Node(arr);
        System.out.print("Original array: ");
        print(head);
        System.out.println("\nk = 3:  " + getValueIterative(head, 3));
        System.out.println("k = 20: " + getValueIterative(head, 20));
    }

}
